Blog post about AiiDA's ORM architecture (#101)
aiidateam/aiida-website: docs/news/posts/2025-09-05-orm.md

---
blogpost: true
category: Blog
tags: architecture
author: Julian Geiger
date: 2025-09-05
---

# Understanding AiiDA's ORM architecture

In this post, we'll take a deep dive into the implementation of AiiDA's Object-Relational Mapping (ORM) system.
We'll explore how AiiDA leverages [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy) to create a flexible (ORM) backend that abstracts database operations and separates concerns between the user-facing Python objects and the underlying database persistence.

## Architecture overview: The multi-layer design

AiiDA's ORM follows a four-layer architecture that provides clean separation between the user interface, business logic, and data persistence:

```
┌─────────────────────┐
│   User interface    │  ← Node (Python ORM class)
│     (orm/nodes)     │
├─────────────────────┤

> │  Backend interface  │  ← BackendNode (Abstract base class)

Reviewer: I don't full understand it but backend interface is not a good name since you also sometimes refer to the database as backend. Is this like a `orm node backend`?

│  (implementation)   │
├─────────────────────┤
│  Database backends  │  ← SqlaNode (SQLAlchemy implementation)
│  (psql_dos/sqlite)  │
├─────────────────────┤
│   Database layer    │  ← DBNode (SQLAlchemy models)
│      (models)       │
└─────────────────────┘
```

Each layer serves a distinct purposes:

* __User interface layer__: Provides a clean, Pythonic API that hides database complexity
* __Backend interface layer__: Defines contracts that all database implementations must follow
* __Database implementation layer__: Handles the specifics of different database systems

> * __Database model layer__: Defines the actual table schemas and relationships using SQLAlchemy's declarative approach

Reviewer: you use here different names than in the table thats a bit confusing

Author: You're right, thanks! I made it consistent now.

        """Return an unstored clone of ourselves."""

    @property
    @abc.abstractmethod
    def uuid(self) -> str:
        """Return the node UUID."""

    @property
    @abc.abstractmethod
    def node_type(self) -> str:
        """Return the node type."""

    # ... many more abstract properties and methods
```

This abstract base class (ABC) ensures that regardless of which database backend is used, the same interface is available to higher-level code.
It defines essential properties like `uuid`, `node_type`, `process_type`, and methods for managing attributes, links, and storage operations.

The abstract class serves as a __contract__ - any backend implementation must provide all these methods and properties.
This guarantees that switching between PostgreSQL and SQLite backends won't break user code, since both implementations satisfy the same interface.

## The bridge: `SqlaNode`

The
[`SqlaNode`](https://github.com/aiidateam/aiida-core/blob/313f342f5d28eeba5967fec8196ed6fce393a77a/src/aiida/storage/psql_dos/orm/nodes.py#L30)
class bridges the abstract `BackendNode` interface with the concrete SQLAlchemy models:

```python

> class SqlaNode(entities.SqlaModelEntity[models.DbNode], ExtrasMixin, BackendNode):

Author: Ah, good catch! I will add a note, thanks!

    """SQLA Node backend entity"""

    MODEL_CLASS = models.DbNode
    USER_CLASS = SqlaUser
    COMPUTER_CLASS = SqlaComputer
    LINK_CLASS = models.DbLink

    def __init__(self, backend, node_type, user, computer=None, **kwargs):
        """Construct a new `BackendNode` instance wrapping a new `DbNode` instance."""
        super().__init__(backend)

        arguments = {
            'node_type': node_type,
            'user': user.bare_model,
            'label': kwargs.get('label', ''),
            'description': kwargs.get('description', ''),
        }

        if computer:
            arguments['dbcomputer'] = computer.bare_model

        self._model = sqla_utils.ModelWrapper(self.MODEL_CLASS(**arguments), backend)

    ...

    def store(self, links=None, clean=True):

>     session = self.backend.get_session()

Reviewer: indentation is wrong

This metadata enables complex provenance queries like "find all structures that were relaxed using PBE exchange-correlation functional."
The `ondelete='CASCADE'` on the output relationship ensures that when a node is deleted, all its incoming links are also removed, maintaining referential integrity.

### The QueryBuilder

The `QueryBuilder` is AiiDA's main Python API to retrieve data from the database.
It provides a uniform, backend-agnostic interface:

<!-- TODO: change example -->
```python
# Simple node query
qb = QueryBuilder()
qb.append(Node, filters={'label': {'like': 'calculation%'}})
results = qb.all()

# Complex relationship query
qb = QueryBuilder()
qb.append(StructureData, tag='structure')
qb.append(CalcJobNode, with_incoming='structure', tag='calc')
qb.append(FolderData, with_incoming='calc', project=['*'])
```

The QueryBuilder automatically:

* __Converts ORM Classes__: Maps Node classes to their corresponding database entities
* __Handles Relationships__: Translates with_incoming/with_outgoing to proper SQL joins

> * __Type Filtering_: Automatically adds filters for node types and subtypes

Reviewer: ```suggestion
* __Type Filtering__: Automatically adds filters for node types and subtypes
```

- **API Integration**: Direct compatibility with FastAPI and other modern frameworks
- **Schema Documentation**: Auto-generated OpenAPI schemas for web interfaces

### NodeCollections: The interface pattern

The `NodeCollection` class provides a clean interface for managing collections of nodes:

```python
class NodeCollection(EntityCollection[NodeType], Generic[NodeType]):
    """The collection of nodes."""

    def delete(self, pk: int) -> None:
        """Delete a Node from the collection with the given id"""
        node = self.get(id=pk)

        if node.base.links.get_incoming().all():
            raise exceptions.InvalidOperation(f'cannot delete Node<{node.pk}> because it has incoming links')

        if node.base.links.get_outgoing().all():
            raise exceptions.InvalidOperation(f'cannot delete Node<{node.pk}> because it has outgoing links')

        self._backend.nodes.delete(pk)
```
This collection pattern provides several benefits:

* __Type safety__: Generic typing ensures you get the correct node type back
* __Validation__: Prevents deletion of nodes with existing links to maintain provenance integrity

> * __Backend abstraction__: Hides database-specific operations behind a clean interface

Reviewer: Isn't performance one of the main reasons? to offer bulk operations

Author: Good point, I didn't know that. Added 👍🏽

## Key Takeaways

AiiDA's ORM architecture makes use of several important design principles:

> 1. **Multi-Layer Abstraction**: The multi-layer design (User Interface → Backend Interface → Database Implementation) provides clean separation while allowing backend-specific optimizations.

Reviewer: we never talk about performance in this blog post, only in the key takeaway

2. **Automatic Backend Adaptation**: The PostgreSQL-to-SQLite conversion system shows how to maintain feature parity across different database engines without code duplication.

3. **Strategic Use of JSON**: JSONB columns provide schema flexibility without sacrificing query performance, with automatic adaptation to SQLite's JSON1 extension.

> 4. **Namespace Organization**: The namespace pattern (`node.base.attributes`, `node.base.repository`) keeps APIs clean while providing rich functionality.

Reviewer: ```suggestion
4. **Namespace Organization**: The namespace pattern (`node.base.attributes`, `node.base.repository`) keeps APIs clean by structuring it in nested categories.
```
